### PR TITLE
Additional Feature Selection operators

### PR DESCRIPTION
**Pull request by bartleyn**
Per #45 , this is a first stab at the four additional feature selection operators. 

Things to note:
- Could use some optimization (caching?)
- Could use some toying around with out-of-bounds parameters (e.g., if num_features < 0 in RFE)
- Terminals for both estimators & scoring functions that plug into the new operators 
  - i.e., RFE takes a supervised estimator so I hard-coded in a linear-kernel SVC; likewise Select* takes a scoring function, so I used chi2.
  - I think this will be of particular importance for generalizing this to regression tasks.

### CODE REVIEW

**Comment by rhiever** [comment]
Reviewing this now. Is RFE the one that's quite slow?

I may also drop `dt_feature_selection()` as a part of this upgrade since it's not supported in sklearn, and exporting it is quite ugly. Perhaps I'll contact the sklearn folks about merging a variant of `dt_feature_selection()` into sklearn.

**Comment by bartleyn** [comment]
Yeah RFE's the slowest, probably because I instantiate the estimator every time.

File: tpot/tpot.py [hunk 1]
```diff
@@ -407,6 +412,11 @@ def export(self, output_file_name):
 from sklearn.cross_validation import StratifiedShuffleSplit
 '''
         if '_dt_feature_selection' in operators_used: pipeline_text += 'from itertools import combinations\n'
+        if '_variance_threshold' in operators_used: pipeline_text += 'from sklearn.feature_selection import VarianceThreshold'
+        if '_select_kbest' in operators_used: pipeline_text += 'from sklearn.feature_selection import SelectKBest'
+        if '_select_percentile' in operators_used: pipeline_text += 'from sklearn.feature_selection import SelectPercentile'
+        if '_select_percentile' or '_select_kbest' in operators_used: pipeline_text += 'from sklearn.feature_selection import chi2'
+        if '_rfe' in operators_used: pipeline_text += 'from sklearn.feature_selection import RFE'
```
Comment (rhiever) [inline tpot/tpot.py]: These imports need newlines `\n` at the end of them.

File: tpot/tpot.py [hunk 2]
```diff
@@ -577,7 +587,46 @@ def export(self, output_file_name):
...
+{2} = {0}[mask + ['class']]
+'''.format(operator[2], operator[3], result_name)
+            elif operator_name == '_select_percentile':
+                operator_text += '''
+#Using Scikit-learn's SelectPercentile for feature selection
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
+
+selector = SelectPercentile(chi2, k={1})
+selector.fit(training_features.values)
+mask = selector.get_support(True)
```
Comment (rhiever) [inline tpot/tpot.py]: Looks like this export case is incomplete and should read something like:

```python
mask_cols = list(training_features[mask].columns) + ['class']
{2} = {0}[mask_cols + ['class']]
```

File: tpot/tpot.py [hunk 3]
```diff
@@ -577,7 +587,46 @@ def export(self, output_file_name):
...
+{2} = {0}[mask + ['class']]
+'''.format(operator[2], operator[3], result_name)
```
Comment (rhiever) [inline tpot/tpot.py]: Looks like this export case is incomplete and should read something like:

```python
mask_cols = list(training_features[mask].columns) + ['class']
{2} = {0}[mask_cols + ['class']]
```

File: tpot/tpot.py [hunk 4]
```diff
@@ -940,6 +989,145 @@ def _subset_df(input_df, start, stop):
...
+        training_class_vals = input_df.loc[input_df['group'] == 'training', 'class'].values
+
+        selector = VarianceThreshold(threshold=threshold)
+        try:
+            selector.fit(training_features) 
+        except ValueError:
+            return input_df.copy()
+        mask = selector.get_support(True)
+        mask_cols = list(training_features[mask].columns) + ['guess', 'class', 'group']
+        #return input_df[[mask , 'guess', 'class', 'group']].copy()
```
Comment (rhiever) [inline tpot/tpot.py]: Please remove commented-out code. :-)

File: tpot/tpot.py [hunk 5]
```diff
@@ -940,6 +989,145 @@ def _subset_df(input_df, start, stop):
...
+        
+        if percentile < 0: 
+            percentile = 0
+        if percentile > 100:
+            percentile = 100
+        if len(training_features.columns) == 0:
+            return input_df.copy()
+
+        selector = SelectPercentile(chi2, percentile=percentile)
+        selector.fit(training_features, training_class_vals)#.reshape((training_class_vals.shape[0], 1)))
```
Comment (rhiever) [inline tpot/tpot.py]: Please remove commented-out code. :-)

File: tpot/tpot.py [hunk 6]
```diff
@@ -940,6 +989,145 @@ def _subset_df(input_df, start, stop):
...
+        training_features = input_df.loc[input_df['group'] == 'training'].drop(['class', 'group', 'guess'], axis=1)
+        training_class_vals = input_df.loc[input_df['group'] == 'training', 'class'].values
+        
+        if k <= 0 or k >= len(training_features.columns):
+            k = 'all'
+        if len(training_features.columns) == 0:
+            return input_df.copy()
+
+        selector = SelectKBest(chi2, k=k)
+        selector.fit(training_features, training_class_vals)#.reshape((training_class_vals.shape[0], 1)))
```
Comment (rhiever) [inline tpot/tpot.py]: Please remove commented-out code. :-)

File: tpot/tpot.py [hunk 7]
```diff
@@ -940,6 +989,145 @@ def _subset_df(input_df, start, stop):
...
+        if num_features < 0:
+            num_features = 1
+        if num_features > len(training_features.columns):
+            num_features = len(training_features.columns)
+        if len(training_features.columns) == 0:
+            return input_df.copy()
+
+        estimator = SVC(kernel='linear')
+        selector = RFE(estimator, n_features_to_select=num_features, step=step)
+        selector.fit(training_features, training_class_vals)#.reshape((training_class_vals.shape[0], 1)))
```
Comment (rhiever) [inline tpot/tpot.py]: Please remove commented-out code. :-)

File: tpot/tpot.py [hunk 8]
```diff
@@ -577,7 +587,46 @@ def export(self, output_file_name):
...
+'''.format(operator[2], operator[3], result_name)
+            elif operator_name == '_select_kbest':
+                operator_text += '''
+#Using Scikit-learn's SelectKBest for feature selection
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
+
+selector = SelectKBest(chi2, k={1})
+selector.fit(training_features.values)
+mask = selector.get_support(True)
+{2} = {0}[mask + ['class']]
```
Comment (rhiever) [inline tpot/tpot.py]: Need to fix the masking here as well.

File: tpot/tpot.py [hunk 9]
```diff
@@ -577,7 +587,46 @@ def export(self, output_file_name):
...
+'''.format(operator[2], operator[3], result_name)
+            elif operator_name == '_select_kbest':
+                operator_text += '''
+#Using Scikit-learn's SelectKBest for feature selection
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
+
+selector = SelectKBest(chi2, k={1})
+selector.fit(training_features.values)
```
Comment (rhiever) [inline tpot/tpot.py]: Need to pass the training classes to `fit()` here.

File: tpot/tpot.py [hunk 10]
```diff
@@ -577,7 +587,46 @@ def export(self, output_file_name):
...
 '''.format(operator[2], operator[3], result_name)
+            elif operator_name == '_variance_threshold':
+                operator_text += '''
+#Using Scikit-learn's VarianceThreshold for feature selection
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
+
+selector = VarianceThreshold(threshold={1})
+selector.fit(training_features.values)
+mask = selector.get_support(True)
+{2} = {0}[mask + ['class']]
```
Comment (rhiever) [inline tpot/tpot.py]: Need to fix the masking here as well.

File: tpot/tpot.py [hunk 11]
```diff
@@ -577,7 +587,46 @@ def export(self, output_file_name):
...
+'''.format(operator[2], operator[3], result_name)
+            elif operator_name == '_select_percentile':
+                operator_text += '''
+#Using Scikit-learn's SelectPercentile for feature selection
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
+
+selector = SelectPercentile(chi2, k={1})
```
Comment (rhiever) [inline tpot/tpot.py]: The parameter should be `percentile`, not `k`.

File: tpot/tpot.py [hunk 12]
```diff
@@ -577,7 +587,46 @@ def export(self, output_file_name):
...
+selector.fit(training_features.values)
```
Comment (rhiever) [inline tpot/tpot.py]: Need to pass the training classes here as well.

File: tpot/tpot.py (hunk 1 above)
Comment (rhiever) [inline tpot/tpot.py]: `_rfe` should also import SVC.

File: tpot/tpot.py [hunk 13]
```diff
@@ -940,6 +989,145 @@ def _subset_df(input_df, start, stop):
...
+        subsetted_df: pandas.DataFrame {n_samples, n_filtered_features + ['guess', 'group', 'class']}
+            Returns a DataFrame containing the the num_pairs best feature pairs
+
+        """
+        
+        training_features = input_df.loc[input_df['group'] == 'training'].drop(['class', 'group', 'guess'], axis=1)
+        training_class_vals = input_df.loc[input_df['group'] == 'training', 'class'].values
+        
+        if step <= 0: 
+            step = 0.1
```
Comment (rhiever) [inline tpot/tpot.py]: Should we cap `step` to a minimum of 0.1, or perhaps 0.05?

File: tpot/tpot.py [hunk 14]
```diff
@@ -577,7 +587,46 @@ def export(self, output_file_name):
...
+mask = selector.get_support(True)
+{2} = {0}[mask + ['class']]
+'''.format(operator[2], operator[3], result_name)
+            elif operator_name == '_rfe':
+                operator_text += '''
+#Using Scikit-learn's Recursive Feature Elimination 
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
 
+selector = RFE(SVC(kernel='linear'), {1}, {2})
+selector.fit(training_features.values)
```
Comment (rhiever) [inline tpot/tpot.py]: Need to pass the training classes here as well.

File: tpot/tpot.py [hunk 15]
```diff
@@ -577,7 +587,46 @@ def export(self, output_file_name):
...
+mask = selector.get_support(True)
+{3} = {0}[mask + ['class']]
```
Comment (rhiever) [inline tpot/tpot.py]: Need to fix the masking here as well.

**Comment by rhiever** [comment]
In addition to the comments I've made directly inline, as you mentioned the exports need to ensure that the variables are within their proper limits.

File: tpot/tpot.py [hunk 16]
```diff
@@ -940,6 +989,145 @@ def _subset_df(input_df, start, stop):
...
+        """
+
+        training_features = input_df.loc[input_df['group'] == 'training'].drop(['class', 'group', 'guess'], axis=1)
+        training_class_vals = input_df.loc[input_df['group'] == 'training', 'class'].values
+
+        selector = VarianceThreshold(threshold=threshold)
+        try:
+            selector.fit(training_features) 
+        except ValueError:
+            return input_df.copy()
```
Comment (rhiever) [inline tpot/tpot.py]: I'm thinking `_variance_threshold()` should return an "empty" DF (with only `class`, `group`, and `guess`) in the case where none of the columns are above the threshold. Otherwise it may become an issue where these `_variance_threshold()` operators with high thresholds are in the pipeline but not actually doing anything.

**Comment by rhiever** [comment]
Alrighty, I think that's all of the comments for now. In all, very nice work on this PR! Thank you for putting this together. Let me know if you want to hack at these comments, else I can merge the PR and clean it up later this week.

>Could use some optimization (caching?)

At least with large steps (>= 0.1), I found RFE to run pretty quickly on my test data sets. Maybe we could limit the RFE steps to >= 0.05 or so and it'll be fine?

>Could use some toying around with out-of-bounds parameters (e.g., if num_features < 0 in RFE)

Except in the export function, everything looks good to me. A great way to test it is to run TPOT with the new operators for several hundred generations. If it doesn't crash by the end of the run, your code has run the gauntlet and *probably* caught all of the possible parameter edge cases. :-)

>Terminals for both estimators & scoring functions that plug into the new operators

Great idea! I'd imagine we can easily encode various estimators and scoring functions with integer values. It may be tricky to have actual estimators and scoring functions as terminals.

Let's file an issue for this to work on after this is merged.

**Comment by bartleyn** [comment]
Yeah I'm on it. I'll work on it when I get home tonight.

**Comment by rhiever** [comment]
You rock! :+1:

File: tpot/tpot.py [hunk 17]
```diff
@@ -577,7 +587,84 @@ def export(self, output_file_name):
...
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
+
+selector = VarianceThreshold(threshold={1})
+try:
+    selector.fit(training_features.values)
+except ValueError:
+    {2} = {0}[['guess', 'class', 'group']]
+try:
+    mask = selector.get_support(True)
+    mask_cols = list(training_features[mask].columns) + ['guess', 'class', 'group']
```
Comment (rhiever) [inline tpot/tpot.py]: There is no `guess` and `group` in the export code.

File: tpot/tpot.py [hunk 18]
```diff
@@ -577,7 +587,84 @@ def export(self, output_file_name):
...
+            elif operator_name == '_variance_threshold':
+                operator_text += '''
+#Using Scikit-learn's VarianceThreshold for feature selection
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
+
+selector = VarianceThreshold(threshold={1})
+try:
+    selector.fit(training_features.values)
+except ValueError:
+    {2} = {0}[['guess', 'class', 'group']]
```
Comment (rhiever) [inline tpot/tpot.py]: There is no `guess` and `group` in the export code.

File: tpot/tpot.py [hunk 19]
```diff
@@ -577,7 +587,84 @@ def export(self, output_file_name):
...
+training_class_vals = {0}.loc[training_indeces, 'class'].values
+if {1} <= 0 or {1} >= len(training_features.columns):
+    {1} = 'all'
+if len(training_features.columns) == 0:
+    {2} = {0}.copy()
+else:
+    selector = SelectKBest(chi2, k={1})
+    selector.fit(training_features.values, training_class_vals)
+    mask = selector.get_support(True)
+    mask_cols = list(training_features[mask].columns) + ['guess', 'class', 'group']
```
Comment (rhiever) [inline tpot/tpot.py]: There is no `guess` and `group` in the export code.

File: tpot/tpot.py [hunk 20]
```diff
@@ -577,7 +587,84 @@ def export(self, output_file_name):
...
+    {1} = 0
+if {1} > 100:
+    {1} = 100
+if len(training_features.columns) == 0:
+    {2} = {0}.copy()
+else:
+    selector = SelectPercentile(chi2, percentile={1})
+    selector.fit(training_features.values, training_class_vals)
+    mask = selector.get_support(True)
+    mask_cols = list(training_features[mask].columns) + ['guess', 'class', 'group']
```
Comment (rhiever) [inline tpot/tpot.py]: There is no `guess` and `group` in the export code.

File: tpot/tpot.py [hunk 21]
```diff
@@ -577,7 +587,84 @@ def export(self, output_file_name):
...
+if {1} > len(training_features.columns):
+    {1} = len(training_features.columns)
+if len(training_features.columns) == 0:
+    {3} = {0}.copy()
+else:
+    selector = RFE(SVC(kernel='linear'), n_features_to_select={1}, step={2})
+    try:
+        selector.fit(training_features.values, training_class_vals)
+        mask = selector.get_support(True)
+        mask_cols = list(training_features[mask].columns) + ['guess', 'class', 'group']
```
Comment (rhiever) [inline tpot/tpot.py]: There is no `guess` and `group` in the export code.

File: tpot/tpot.py [hunk 22]
```diff
@@ -577,7 +587,84 @@ def export(self, output_file_name):
...
+        {3} = {0}[mask_cols]
+    except ValueError:
+        {3} = {0}[['guess', 'class', 'group']]
```
Comment (rhiever) [inline tpot/tpot.py]: There is no `guess` and `group` in the export code.

File: tpot/tpot.py [hunk 23]
```diff
@@ -577,7 +587,84 @@ def export(self, output_file_name):
...
+except:
+    pass
+'''.format(operator[2], operator[3], result_name)
+            elif operator_name == '_select_kbest':
+                operator_text += '''
+#Using Scikit-learn's SelectKBest for feature selection
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
+training_class_vals = {0}.loc[training_indeces, 'class'].values
+if {1} <= 0 or {1} >= len(training_features.columns):
+    {1} = 'all'
```
Comment (rhiever) [inline tpot/tpot.py]: This should be the variable, not the value.

File: tpot/tpot.py [hunk 24]
```diff
@@ -577,7 +587,84 @@ def export(self, output_file_name):
...
+    {2} = {0}[mask_cols]
+'''.format(operator[2], operator[3], result_name)
+            elif operator_name == '_select_percentile':
+                operator_text += '''
+#Using Scikit-learn's SelectPercentile for feature selection
+training_features = {0}.loc[training_indeces].drop('class', axis=1)
+training_class_vals = {0}.loc[training_indeces, 'class'].values
 
+if {1} < 0: 
+    {1} = 0
```
Comment (rhiever) [inline tpot/tpot.py]: This should be the variable `percentile`, not the value.

File: tpot/tpot.py [hunk 25]
```diff
@@ -577,7 +587,84 @@ def export(self, output_file_name):
...
+if {1} > 100:
+    {1} = 100
```
Comment (rhiever) [inline tpot/tpot.py]: This should be the variable `percentile`, not the value.

File: tpot/tpot.py [hunk 26]
```diff
@@ -407,6 +412,11 @@ def export(self, output_file_name):
 from sklearn.cross_validation import StratifiedShuffleSplit
 '''
         if '_dt_feature_selection' in operators_used: pipeline_text += 'from itertools import combinations\n'
+        if '_variance_threshold' in operators_used: pipeline_text += 'from sklearn.feature_selection import VarianceThreshold\n'
+        if '_select_kbest' in operators_used: pipeline_text += 'from sklearn.feature_selection import SelectKBest\n'
+        if '_select_percentile' in operators_used: pipeline_text += 'from sklearn.feature_selection import SelectPercentile\n'
+        if '_select_percentile' or '_select_kbest' in operators_used: pipeline_text += 'from sklearn.feature_selection import chi2\n'
+        if '_rfe' in operators_used: pipeline_text += 'from sklearn.feature_selection import RFE\n'
```
Comment (rhiever) [inline tpot/tpot.py]: RFE should also import SVC.

**Comment by rhiever** [comment]
Just noting some minor issues in the export code. About to do some final tests of the pipeline operators themselves, and if those turn out fine, I'll merge this and do some final cleanup.

**Comment by bartleyn** [comment]
Gotcha, thanks! I won't concern myself with cleaning up the export code then. I'll get it right one day though, I promise :P.

**Comment by rhiever** [comment]
Looks good! Thanks again for your PR! :-)
